src/sentry/rules/history/preview.py

```diff
@@ -387,68 +387,92 @@ def apply_frequency_conditions(
     """
     Applies frequency conditions to issue state activity.
     """
-    # TODO: separate the conditions so we can reuse queries for conditions of the same class
-    conditions = []
+    condition_types = defaultdict(list)
     for condition_data in frequency_conditions:
         condition_cls = rules.get(condition_data["id"])
         if condition_cls is None:
             raise PreviewException
-        conditions.append(condition_cls(project, data=condition_data))
+        condition_types[condition_data["id"]].append(condition_cls(project, data=condition_data))
 
+    filtered_activity = defaultdict(list)
     if condition_match == "all":
-        filtered_activity = {}
         for group, activities in group_activity.items():
-            # TODO: only EventFrequencyCondition is supported right now, so we can reuse the same query
-            buckets = get_frequency_buckets(
-                project, start, end, group, dataset_map.get(group, None)
-            )
-            if not has_issue_state_condition:
-                # If there are no issue state change conditions, then we won't have any initial activities to base our
-                # frequency condition queries off of. Instead, we take the first frequency condition and create the
-                # initial activities from that
-                for bucket_time in buckets.keys():
-                    activity = ConditionActivity(
+            init_activities_from_freq_cond = False
+            for conditions in condition_types.values():
```

Comment:
Now that we support more than one type of freq condition, we can batch them up by id and reuse the same frequency buckets for the all freq conditions of that id